whclient/client.go

```diff
@@ -0,0 +1,130 @@
+package whclient
+
+import (
+	"math/rand"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/taskcluster/webhooktunnel/wsmux"
+)
+
+// Client is used to connect to a proxy and serve endpoints
+// defined in Handler
+type Client struct {
+	ID string
+	// handler should be a mux to handle different end points
+	Handler   http.Handler
+	Config    wsmux.Config
+	ProxyAddr string // Address of proxy server for connection
+	Retry     RetryConfig
+}
+
+const (
+	defaultInitialInterval     = 500 * time.Millisecond
+	defaultMaxInterval         = 60 * time.Second
+	defaultMaxElapsedTime      = 3 * time.Minute
```

Comment:
I found this approach in github.com/cenkalti/backoff.git . It's used in the tc http backoff library.